### ISSUE TITLE
Mintty personalization: collecting themes and sounds

### ISSUE THREAD

**Issue by mintty** [mintty/mintty#711]
This issue calls for contribution of ANSI colour schemes (“themes”)
and bell sounds for personal adaptation of mintty appearance.

To contribute a theme or sound, please add the following information here:
* link to original source
* link to your modified version if applicable
* link to license or other permission of redistribution of original source

**Comment by mintty** [comment]
Released 2.8.2 with some themes included.

**Comment by naknak-** [comment]
There are many public domain sound effects on Freesound.  Not everything is public domain (aka "[Creative Commons 0](https://creativecommons.org/publicdomain/zero/1.0/)") but you can search by license type. 

For example [CC:0 license, 300 ms or shorter](https://freesound.org/search/?q=&s=num_downloads+desc&g=1&advanced=1&f=license%3A%22Creative+Commons+0%22+duration%3A%5B0.05+TO+0.3%5D#sound)

A few that I use:
- [Pop, 45ms](https://freesound.org/people/greenvwbeetle/sounds/328117/)
- [8bit Question Mark, 111ms](https://freesound.org/people/plasterbrain/sounds/399093/)
- [Retro video game sfx - Move, 105ms](https://freesound.org/people/OwlStorm/sounds/404740/)
- [Metallic ping, 110ms](https://freesound.org/people/high_festiva/sounds/194439/)

**Comment by mintty** [comment]
When I browsed some sound sites a while ago, I considered very few sounds suitable, also considering quality (e.g. they should have no disturbing click when they begin or end), and also I found most sites quite deterring when they call their offer "free" but require you to login.
Also the purpose of this issue was to distribute some of the burden to investigate into good and free sounds to the community. So if you think you've found some good ones, please upload well-tuned .wav files here, attached with a statement of your findings about their license.

**Comment by naknak-** [comment]
Oh right - I forgot to mention the login is not actually required.  The links are just hidden, revealed by CSS in Stylus or some similar extension: `.player .metadata { display:inline-block; font-size:40%; }`

The spectrogram that's attached to each one would clearly show defects like clicks, though that isn't common at all (if it's common for you, and you're converting them yourself, it may be a transcoding artifact - try ffmpeg).

What *is* somewhat common is delayed start (e.g., 30ms of leading silence) - undesirable for UI-feedback purposes, and easily avoided because it shows up on spectrogram.  

As to tuning, like volume normalization, that's something that has to be done to some common standard.  Given 10 different choices of sounds, it's important that they all be about the same volume.

I use different bells for different sessions and my own criteria are straightforward - under 150ms, with principal audio fitting into the first ~50ms, no opening silence, not scratchy or shrill.  Given those criteria, over half of the search results are suitable, and about half of the rejects can be rejected on sight.  

If you're finding very few to be suitable, you may have very different criteria.

**Comment by mintty** [comment]
I appreciate your approach but your hints somehow miss the point. This all costs a lot of time, browsing the site, selecting suitable sounds, fiddling out hidden links, getting ffmpeg to run (which is unfortunately not a cygwin package and fails to build out-of-the-box on cygwin), and tuning sounds.
As you've apparently done this already, why don't you simply upload your selection, along with links to the original pages where the licenses can be checked?

**Comment by BrianInglis** [comment]
Pointers in https://opensource.com/article/18/6/sound-themes-linux

**Comment by naknak-** [comment]
I can certainly upload files.  I want to understand the goal.  To me this is a solved problem; to you it's not.  That means we don't see the problem the same way.  You're saying very few sounds are suitable.  That makes me think I am missing some implicit requirement.  

If I pick out 20 distinct sounds, download, convert, normalize them to common volume, and upload here and you say "these are all no good,"  that is of no help to anyone. 

I posted a few examples of what I'd consider suitable, along with a link to the license they all share (CC:0, also linked on the destination pages).  You can click the links and click play, no login or CSS or ffmpeg required.  If those suit you, that's something contributors can work with.  If they don't, you can describe how they differ from what you want, and that's also something contributors can work with.

**Comment by mintty** [comment]
I've checked your 4 mentioned sounds, one of them is a bit sharp in my ear but OK. They are all similar, let's call it the "pop" style. I would accept them in a collection with also some other styles.

Personally, I had found a few nice ones, e.g. cartoon008.wav on the page http://www.grsites.com/archive/sounds/category/23/, which I'd like to include. But their Terms of Use allow "personal use" of the sounds which is fuzzy; is bundling and distributing with open source software still "personal use"? I've sent them a request.

**Comment by BrianInglis** [comment]
GRsites sell sets of all the sounds for either personal or commercial use. 
We should presume that these are not free to distribute, modify, provide for a service charge, etc. so should not bundle them in any other package, and possibly not provide them on github, even if you can get favourable licence terms for personal use, as they may prevent files being included, or packages being accepted, for distribution under the terms required by the distributors. 
You should consult with your downstream distributors about personal licences, licence terms, and the licences and terms under which you provide contributions to downstream distributions. 
Your Cygwin docs reference but do not include LICENCE.GPL or LICENCE.bundling and this should be remedied in your next release of the package cygport src_install function docinto/dodoc actions.

**Comment by mintty** [comment]
Thanks, Brian, for the comment about distribution problems with GR sounds. That's exactly the kind of trouble that I don't want to care about and why I haven't bundled any sounds so far although I'd like to.
Also, thanks for the hint about missing files in /usr/share/doc/mintty, I'll fix that; dodoc LICENSE* should work.

**Comment by naknak-** [comment]
179 short sounds: [bells179.zip](https://github.com/mintty/mintty/files/3078034/bells179.zip) (1.8MB); [[manifest](https://github.com/mintty/mintty/files/3078033/MANIFEST-179.txt)]

They are all:

- licensed under [CC0 (public domain)](https://creativecommons.org/publicdomain/zero/1.0/): "_copy, modify, distribute and perform the work, even for commercial purposes, all without asking permission._ "
- evocatively named
- between 48ms and 300ms long (most under 200ms)
- auditioned on headphones and speakers both
- normalized to sound about equally loud
- mono 16-bit PCM WAV at 44100 Hz.  This is what Microsoft distributes, except theirs are stereo
- free of pops, clipping or audible noise floors 
- free of leading silence
- free of bad data according to sox and FFmpeg
- tagged consistently in the RIFF INFO block: 
	* artist: source URL
	* title: brief description (also in filename)
	* genre: category (also in filename)
	* license: "CC0 (public domain)" plus a link to the license text
	* comment: the first five tags provided by the author
	* date: if provided by the author, otherwise blank
- catalogued with other metadata in the included `_MANIFEST.yaml`
- reasonably distinct; many sounds have high and low-pitched variants to accommodate differences in hearing sensitivity, speakers and taste
- categorized:
	* vocal: humans or animals
	* percussive: things hitting things
	* resonant: strings, whistles, jugs, tuning forks
	* synthetic: chipsounds, tone sweeps, laser guns
	* ambient
	
---------------
Notes that could be useful to others:

[WaveGain](http://www.rarewares.org/others.php#wavegain) gave good normalization results.  FFmpeg's volume normalization gave bad results at any setting I tried: one file would be barely audible and the next would be too loud.  

Foobar2000 is good for this kind of work.  Importantly, playlist autoadvance can be disabled and anything can be bound to a hotkey.

A WAV's INFO block can hold anything but only some fields are EXIF standard.  The useful standard fields are:

    field   description     ffmpeg alias (actual field name works too)
    -----   -----------     ------------
	IARL 	ArchivalLocation 	
	IART 	Artist 		"artist" 
	ICMT 	Comment 	"comment" 
	ICOP 	Copyright	"copyright"
	ICRD 	DateCreated 	"date"
	IENG 	Engineer 	
	IGNR 	Genre		"genre"
	ILNG	Language	"language"
	INAM 	Title 		"title"
	IPRD 	Product 	"album"
	ISFT    Software        "encoder" - autofilled by libavformat	
	ITCH 	Technician 	"encoded_by" 	 

Fuller details at [RIFF tag specs](https://sno.phy.queensu.ca/~phil/exiftool/TagNames/RIFF.html) and [FFmpeg AVI tags](https://wiki.multimedia.cx/index.php/FFmpeg_Metadata#AVI)

---------------

Some FFmpeg recipes (in bash):

1. bulk downmix to mono, output to ../mono/ 
```
for x in $( exiftool -p '$filename' -if '$numchannels>1' -q -r *wav ) ; do ffmpeg -v error -i "$x" -ac 1 ../mono/"$x"  ; done
cp -n *wav ../mono # already-mono files skipped in step 1
```
    
2. apply tag to one $file
```
ffmpeg -v error -i $file \
    -metadata comment='...'\
    -metadata artist='...'\
    -metadata album= \
    -metadata track= \
    -metadata genre= \
    -metadata encoded_by= \
    -metadata IENG= \
    -metadata copyright='CC0 (public domain) creativecommons.org/publicdomain/zero/1.0' \
  tagged/$file &
```

3. apply tags from same-named file in other directory to file in this directory, output to ../retagged/
```
ffmpeg -v error -i ../tagged/"$wav" -i "$wav" -map 1 -c copy -map_metadata 0 -map_metadata:s:a 0:s:a ../retagged/"$wav"
```

4. trim leading silence with -60dB threshold, output to ../trim/
```
ffmpeg -v error -i "$wav" -af silenceremove=1:0:-60dB:window=5 "trim/$wav"
```
    
5. get tags, duration, etc.  
```
mediainfo --Inform="Audio;%Duration%" $wav  # length in ms
soxi -D "$wav" # length in seconds
ffprobe -v error -show_format "$wav" |grep ^TAG:  # ffprobe comes with ffmpeg
comment=$( ffprobe -v error -show_format "$wav" |grep -Po "^TAG:comment=\K.*" ) # capture the value of one tag
man exiftool 
```
	
6. sixel waveform to terminal, for trial-and-error sessions with silence trim or normalization
```
die() { echo "$@" >&2; exit 1; }
for wav in "$@"; do
  [[ $wav == *.wav ]] || die "$wav not a wav"
  [[ -f $wav ]] || die "$wav doesn't exist"
  png=${wav%.wav}.png
  rm -f "$png"
  printf -v dur "%.3fs" $( soxi -D "$wav" )
  sox "$wav" -t dat - | tail -n +3 | gnuplot -e "
	set term png size 320,120;
	set output '$png';
	set yr [-1:1];
	unset key;
	unset tics;
	unset border;
	set lmargin 0.0001;
	set rmargin 0;
	set tmargin 0;
	set bmargin 0;
	set termoption font ',9';
	set obj 1 rectangle behind from screen 0,0 to screen 1,1;
	set obj 1 fillstyle solid 1.0 fillcolor rgbcolor '#dddddd';
	set title '$wav
	$dur' offset 0,-3 noenhanced  textcolor 'black';
	plot '-' with lines lt rgb '#7788bb';
  "
  img2sixel "$png";#  du -Bk "$wav"
done
```

**Comment by mintty** [comment]
That is an amazing well-done contribution, thank you!
The only problem still is that a list of 179 sounds in the menu would be quite long and not offer a concise choice. Therefore my proposed plan ist:
* Either distribute all 179 with the package but not deploy all by default (or offer an optional download suggestion).
* Add a helper script `minsound` (like `mintheme`, or mintheme with a sound parameter) to support selective deployment of available sounds into the menu.
* Make a selection of them for implicit pre-deployment in the menu; I've now selected the 69 I like most but I guess that may still be too many. I'm attaching a list of them, you may comment, or select, or/and add a few of your own preference.
[bells69.txt](https://github.com/mintty/mintty/files/3086450/bells69.txt)

**Comment by naknak-** [comment]
Removing is hard!  I started with over 600 and once the obvious losers are gone it's a slow process to choose the best.    

What about having them in a directory?  The Wave input box is there in Terminal menu already, could that be a `…` browse button that opened a directory like "/usr/share/mintty/bells"?  Choosing might still be tedious (browse, select, play, repeat) but the user will see the sounds all in one place and can use his media player to check them out quickly, if the names are not enough.  This looks to be a common method that other packages use to provide aesthetic alternatives (e..g, /usr/share/highlight/themes or /usr/share/vim/vim80/colors).

Your choices look fine; we seem to both agree that "resonant" is the strongest category overall.  I'd say ambient-swoosh sounds better than ambient-broom.   Of the pet sounds, vocal-yip is cleaner than vocal-meow (meow has some background noise), and of the low-pitched vocals vocal-uh is cleaner than vocal-oof.  Apart from those, I see all my own favorites on your list.  

If I were to make a list of just 10, like for a dropdown menu, I'd suggest:

- ambient-drip-lo
- ambient-glass
- resonant-organ
- resonant-twang, 
- resonant-reverb
- synthetic-spike
- synthetic-rave (or harm or danger, which are approximately higher pitch versions of the same thing)
- a cute vocal sound (hah or boo)
- a bass sound (synthkick-lo, thud, tom-lo, bass)
- and one ~60ms fast sound (swoosh, pipe, knock, bleep, drop)

**Comment by mintty** [comment]
Released 3.0.1.
Closing this issue as I think there's now a sufficient repository of bell sounds included.
Further contributions will still be considered, of course.
